### PR TITLE
[F2F-663] - CIC | Platform | Add API Tests to Test Container Script

### PR DESCRIPTION
**Pull request by mudassar-jafer**
### What changed

- Updated run-tests.sh run the end to end tests for CIC API.
- Updated run-tests-local.sh scripts to create an image and run tests locally. This allows creation of the container and verification of run-tests.sh 
- Changed docker image to Ubuntu to allow support for node 18
- Added jest-junit reporter for generating test reports.

### Why did it change

To run end to end tests in the code pipeline.

### Issue tracking
https://govukverify.atlassian.net/browse/F2F-663

- [F2F-XXX](https://govukverify.atlassian.net/browse/F2F-XXX)

### Environment variables or secrets

- [x] No environment variables or secrets were added or changed

- [x] Documented in the [README](./blob/main/README.md)

### Other considerations

- [x] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: run-tests.sh [hunk 1]
```diff
@@ -1,73 +1,21 @@
...
-status_code="$(curl --silent --output /dev/null --write-out '%{http_code}' "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200")"
+#The CFN variables seem to include quotes when used in tests these must be removed before assigning them.
+CFN_CICBackendURL_NoQuotes=$(remove_quotes $CFN_CICBackendURL)
+export DEV_CRI_CIC_API_URL=`echo ${CFN_CICBackendURL_NoQuotes%/}`
+export DEV_IPV_STUB_URL="https://erveje5km8.execute-api.eu-west-2.amazonaws.com/dev/start"
+
+cd ./src; npm run test:e2e
+error_code=$?
+
+cp -rf $TEST_REPORT_DIR $TEST_REPORT_ABSOLUTE_DIR
```
Comment (darrendunford) [inline run-tests.sh:19]:
```suggestion
cp -rf "$TEST_REPORT_DIR" "$TEST_REPORT_ABSOLUTE_DIR"
```
prevent unexpected before from word splitting if the env vars contain spaces

File: run-tests.sh [hunk 2]
```diff
@@ -1,73 +1,21 @@
...
+remove_quotes () {
+  echo "$1" | tr -d '"'
+}
+
+declare error_code
 # shellcheck disable=SC2154
-# status_code="$(curl --silent --output /dev/null --write-out '%{http_code}' "$CFN_HelloWorldApi")"
-status_code="$(curl --silent --output /dev/null --write-out '%{http_code}' "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200")"
+#The CFN variables seem to include quotes when used in tests these must be removed before assigning them.
+CFN_CICBackendURL_NoQuotes=$(remove_quotes $CFN_CICBackendURL)
```
Comment (darrendunford) [inline run-tests.sh:12]:
```suggestion
CFN_CICBackendURL_NoQuotes=$(remove_quotes "$CFN_CICBackendURL")
```
my linter is complaining about this!

File: run-tests.sh (hunk 1 above)
Comment (darrendunford) [inline run-tests.sh:19]: does `TEST_REPORT_ABSOLUTE_DIR` need to be created before `cp`ing files in to it?

File: run-tests.sh [hunk 3]
```diff
@@ -1,73 +1,21 @@
...
+export DEV_CRI_CIC_API_URL=`echo ${CFN_CICBackendURL_NoQuotes%/}`
+export DEV_IPV_STUB_URL="https://erveje5km8.execute-api.eu-west-2.amazonaws.com/dev/start"
```
[the remark this answers is not in the document]
Comment (mudassar-jafer) [reply run-tests.sh]: I am now using cic-ipv-stub-IPVStubExecuteUrl. You should see this value being imported in the template.

File: Dockerfile [hunk 4]
```diff
@@ -1,16 +1,14 @@
-FROM amazonlinux:2
+FROM ubuntu:latest
```
Comment (richa-misra) [inline Dockerfile:1]: did we want to pin to a version or SHA?

File: run-tests.sh [hunk 5]
```diff
@@ -1,73 +1,19 @@
...
+  echo "$1" | tr -d '"'
+}
+
+declare error_code
 # shellcheck disable=SC2154
-# status_code="$(curl --silent --output /dev/null --write-out '%{http_code}' "$CFN_HelloWorldApi")"
-status_code="$(curl --silent --output /dev/null --write-out '%{http_code}' "https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200")"
+#The CFN variables seem to include quotes when used in tests these must be removed before assigning them.
+CFN_CICBackendURL_NoQuotes=$(remove_quotes "$CFN_CICBackendURL")
+export DEV_CRI_CIC_API_URL=`echo ${CFN_CICBackendURL_NoQuotes%/}`
```
Comment (richa-misra) [inline run-tests.sh:13]: will these github secrets: https://github.com/alphagov/di-ipv-cri-cic-api/blob/13e487a7e20da446839a748ea85de6de081d4ae9/.github/workflows/pull-request.yml#LL25C1-L27C52 still needed after this change ?

File: .vscode/launch.json
[the remark this answers is not in the document]
Comment (mudassar-jafer) [reply .vscode/launch.json:1]: This is for setting up VS code correctly for tests. I committed it so it is available to anyone who wants to use vscode.

File: .vscode/settings.json
[the remark this answers is not in the document]
Comment (mudassar-jafer) [reply .vscode/settings.json:1]: same as previous comment. For vscode setup.

File: Dockerfile (hunk 4 above)
Comment (mudassar-jafer) [reply Dockerfile:1]: I will change it to 20.10.7 which is the current latest.

File: run-tests.sh (hunk 5 above)
Comment (mudassar-jafer) [reply run-tests.sh:13]: Yes, this script is run inside the container and this is the only way I could get the values by using the outputs.

**Review by darrendunford — APPROVED** [review]

File: Dockerfile (hunk 4 above)
Comment (mudassar-jafer) [reply Dockerfile:1]: I can only get ubuntu latests otherwise it seems to fail for various reasons.

**Review by richa-misra — APPROVED** [review]

**Pull request merged**